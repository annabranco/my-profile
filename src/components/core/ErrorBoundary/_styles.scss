
.error-boundary__title {
  margin-left : 0;
  width       : auto;
  font-family : $font__subtitle;
  font-size   : 2rem;
  text-align  : left;
}

.error-boundary__notify--wrapper {
	position        : relative;
	top             : 50%;
	display         : flex;
	flex-direction   : column;
	justify-content : center;
	align-items     : center;
	width           : 80vw;
}

.error-boundary__sorry-msg {
  margin      : 0 0 20px;
  width       : 80%;
  font-style  : italic;
  font-family : $font__subtitle;
  font-size   : 1.3rem;
  line-height : 1.5;
  color       : $blue--light;
  text-align  : left;
}

.error-boundary__notify-button {
  position    : relative;
  right       : 0;
  margin      : 4px 0 0 5px;
  font-family : $font__title-alt;
  font-size   : 1.3rem;
  color       : $black;
	}

.error-boundary__notify-button--text {
  display        : inline;
  margin-left    : 15px;
  vertical-align : middle;
  text-shadow    : 0 0 1px $black;
  font-size      : 2rem;
}