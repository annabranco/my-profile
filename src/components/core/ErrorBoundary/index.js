import React, { Component } from 'react';
import { PropTypes } from 'prop-types';

class ErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.node
  };

  static defaultProps = {
    children: undefined
  };

  state = { hasError: false };

  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI.
    return { hasError: error };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
