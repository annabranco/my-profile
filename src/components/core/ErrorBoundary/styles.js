import styled, { css } from 'styled-components';
import {
  colorWhite,
  colorBlueLight,
  colorBlack,
  colorGrayNormal,
  fontSubtitle,
  fontTitleAlt
} from '../../../styles/theme';
import { FullScreen, MainBackground } from '../../../styles/global';
import { SocialLinkButton } from '../../views/Social/styles';

export const ErrorSection = styled.div`
  ${MainBackground}
  ${FullScreen};
  margin: 0;
`;

export const ErrorTitle = styled.h2`
  margin-left: 0;
  width: auto;
  font-family: ${fontSubtitle};
  font-size: 2rem;
  text-align: left;
`;

export const ErrorNotifyWrapper = styled.div`
  position: relative;
  top: 50%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80vw;
`;

export const ErrorSorryMessage = styled.p`
  margin: 0 0 20px;
  width: 80%;
  font-style: italic;
  font-family: ${fontSubtitle};
  font-size: 1.3rem;
  line-height: 1.5;
  color: ${colorBlueLight};
  text-align: left;
`;

export const ErrorDetailArea = styled.div`
  background: ${colorWhite};
  width: 60vw;
  padding: 10px 40px;
  box-sizing: border-box;
  border-radius: 8px 0;
  margin-top: 15px;
  box-shadow: 0 0 5px 2px ${colorWhite};

  &::before {
    content: 'ERROR:';
    font-size: 0.7rem;
    font-family: ${fontTitleAlt};
    color: ${colorGrayNormal};
  }
`;

export const ErrorDetailText = styled.p`
  font-size: 1rem;
`;

export const ErrorNotifyButton = styled(SocialLinkButton)`
  margin-top: 40px;
  position: fixed;
  bottom: 5%;
  opacity: 0;

  ${({ visible }) =>
    visible &&
    css`
      opacity: 1;
      transition: opacity 10s;
    `}
`;

export const ErrorNotifyButtonText = styled.p`
  position: relative;
  right: 0;
  margin: 4px 0 0 5px;
  font-family: ${fontTitleAlt};
  font-size: 1.3rem;
  color: ${colorBlack};
`;

export const ErrorNotifyButtonIcon = styled.i`
  display: inline;
  margin-left: 15px;
  vertical-align: middle;
  text-shadow: 0 0 1px ${colorBlack};
  font-size: 2rem;
`;
