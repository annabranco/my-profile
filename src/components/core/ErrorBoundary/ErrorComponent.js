import React, { useEffect } from 'react';
import { string } from 'prop-types';
import appInfo from '../../../../package.json';
import { useStateWithLabel } from '../../../utils/hooks';
import { Barquinho } from '../../../assets/images';
import { NOTIFY_BUTTON } from '../../../constants';
import {
  DetailsText,
  ErrorDetailsArea,
  ErrorSection,
  ErrorTitle,
  NotificationArea,
  NotifyButton,
  NotifyButtonIcon,
  NotifyButtonText,
  SorryText
} from './styles';
import { MeuBarquinho } from '../../sections/MyInfoPage/styles';
import { errorTextsPropType } from '../../../types';
import { isDesktop } from '../../../utils/device';

const DEFAULT_ERROR_TEXTS = {
  errorLine1: "I'm awfully sorry but something unexpected had happened. :(",
  errorLine2:
    "I'd really appreciate if you could notify me of this error on my issues page.",
  notifyMe: 'Notify me',
  title: "That's embarassing..."
};

const REPORT_ISSUE_PAGE = `${appInfo.bugs.url}/new`;

const ErrorComponent = ({ error, texts }) => {
  const [showNotifyButton, toggleNotifyButton] = useStateWithLabel(
    false,
    NOTIFY_BUTTON
  );

  useEffect(() => {
    setTimeout(() => toggleNotifyButton(true), 5000);
  });

  return (
    <ErrorSection>
      {isDesktop && (
        <MeuBarquinho src={Barquinho} alt="Navigating beautifully" />
      )}
      <ErrorTitle>{texts.title}</ErrorTitle>
      <NotificationArea>
        <SorryText>{texts.errorLine1}</SorryText>
        <SorryText>{texts.errorLine2}</SorryText>
        <ErrorDetailsArea>
          <DetailsText>{error}</DetailsText>
        </ErrorDetailsArea>
        <NotifyButton
          visible={showNotifyButton}
          href={REPORT_ISSUE_PAGE}
          target="_Blank"
          rel="noopener noreferrer"
        >
          <NotifyButtonText>{texts.notifyMe}</NotifyButtonText>
          <NotifyButtonIcon className="fab fa-github-alt" />
        </NotifyButton>
      </NotificationArea>
    </ErrorSection>
  );
};

ErrorComponent.propTypes = {
  error: string.isRequired,
  texts: errorTextsPropType
};

ErrorComponent.defaultProps = {
  texts: DEFAULT_ERROR_TEXTS
};

export default ErrorComponent;
