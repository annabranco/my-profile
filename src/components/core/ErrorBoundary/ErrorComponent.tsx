import { ReactElement, useEffect } from 'react';
import { string } from 'prop-types';
import { IErrorText } from '../../../types/interfaces';
import appInfo from '../../../../package.json';
import { useStateWithLabel } from '../../../utils/hooks';
import { isDesktop } from '../../../utils/device';
import { Barquinho } from '../../../assets/images';
import { NOTIFY_BUTTON } from '../../../constants';
import { errorTextsPropType } from '../../../types/propTypes';
import { MeuBarquinho } from '../../sections/MyInfoPage/styles';
import {
  DetailsText,
  ErrorDetailsArea,
  ErrorSection,
  ErrorTitle,
  NotificationArea,
  NotifyButton,
  NotifyButtonIcon,
  NotifyButtonText,
  SorryText
} from './styles';

const DEFAULT_ERROR_TEXTS = {
  errorLine1: "I'm awfully sorry but something unexpected had happened. :(",
  errorLine2:
    "I'd really appreciate if you could notify me of this error on my issues page.",
  notifyMe: 'Notify me',
  title: "That's embarassing..."
};

const REPORT_ISSUE_PAGE = `${appInfo.bugs.url}/new`;

interface Props {
  error: string;
  texts: IErrorText;
}

const ErrorComponent = ({ error, texts }: Props): ReactElement => {
  const [showNotifyButton, toggleNotifyButton] = useStateWithLabel<boolean>(
    false,
    NOTIFY_BUTTON
  );

  useEffect(() => {
    setTimeout(() => toggleNotifyButton(true), 5000);
    // toggleNotifyButton(true);
  }, []);

  return (
    <ErrorSection>
      {isDesktop && (
        <MeuBarquinho src={Barquinho} alt="Navigating beautifully" />
      )}
      <ErrorTitle>{texts.title}</ErrorTitle>
      <NotificationArea>
        <SorryText>{texts.errorLine1}</SorryText>
        <SorryText>{texts.errorLine2}</SorryText>
        <ErrorDetailsArea>
          <DetailsText>{error}</DetailsText>
        </ErrorDetailsArea>
        <NotifyButton
          visible={showNotifyButton}
          href={REPORT_ISSUE_PAGE}
          target="_Blank"
          rel="noopener noreferrer"
        >
          <NotifyButtonText>{texts.notifyMe}</NotifyButtonText>
          <NotifyButtonIcon className="fab fa-github-alt" />
        </NotifyButton>
      </NotificationArea>
    </ErrorSection>
  );
};

ErrorComponent.propTypes = {
  error: string.isRequired,
  texts: errorTextsPropType
};

ErrorComponent.defaultProps = {
  texts: DEFAULT_ERROR_TEXTS
};

export default ErrorComponent;
