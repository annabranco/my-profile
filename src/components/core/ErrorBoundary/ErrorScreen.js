import React, { Component } from 'react';
import { string } from 'prop-types';
import { Barquinho } from '../../../images';

import {
  ErrorSection,
  ErrorTitle,
  ErrorNotifyWrapper,
  ErrorSorryMessage,
  ErrorNotifyButton,
  ErrorNotifyButtonText,
  ErrorNotifyButtonIcon,
  ErrorDetailArea,
  ErrorDetailText
} from './styles';
import { MeuBarquinho } from '../../views/MyInfoPage/styles';

const DEFAULT_ERROR_TEXTS = {
  errorLine1: "I'm awfully sorry but something unexpected had happened. :(",
  errorLine2:
    "I'd really appreciate if you could notify me of this error on my issues page.",
  notifyMe: 'Notify me',
  title: "That's embarassing..."
};

class ErrorScreen extends Component {
  static propTypes = {
    error: string.isRequired,
    texts: string
  };

  static defaultProps = {
    texts: DEFAULT_ERROR_TEXTS
  };

  state = { showNotifyButton: false };

  componentDidMount() {
    setTimeout(() => this.toggleNotifyButton(true), 5000);
  }

  toggleNotifyButton = () => this.setState({ showNotifyButton: true });

  render() {
    const { texts, error } = this.props;

    return (
      <ErrorSection>
        <MeuBarquinho src={Barquinho} alt="Navigating beautifully" />
        <ErrorTitle>{texts.title}</ErrorTitle>
        <ErrorNotifyWrapper>
          <ErrorSorryMessage>{texts.errorLine1}</ErrorSorryMessage>
          <ErrorSorryMessage>{texts.errorLine2}</ErrorSorryMessage>
          <ErrorDetailArea>
            <ErrorDetailText>{error}</ErrorDetailText>
          </ErrorDetailArea>
          <ErrorNotifyButton
            visible={this.state.showNotifyButton}
            href="https://github.com/annabranco/my-profile/issues/new"
            target="_Blank"
            rel="noopener noreferrer"
          >
            <ErrorNotifyButtonText>{texts.notifyMe}</ErrorNotifyButtonText>
            <ErrorNotifyButtonIcon className="fab fa-github-alt" />
          </ErrorNotifyButton>
        </ErrorNotifyWrapper>
      </ErrorSection>
    );
  }
}

export default ErrorScreen;
