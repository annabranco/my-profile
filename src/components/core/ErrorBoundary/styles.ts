import styled from 'styled-components';
import {
  colorWhite,
  colorBlueLight,
  colorBlack,
  colorGrayNormal,
  fontSubtitle,
  fontTitleAlt
} from '../../../styles/theme';
import { FullScreen, MainBackground } from '../../../styles/global';
import { LinkButton } from '../../elements/Social/styles';

export const DetailsText = styled.p`
  font-size: 1rem;
`;
DetailsText.displayName = 'DetailsText';

export const ErrorDetailsArea = styled.div`
  background: ${colorWhite};
  width: 100%;
  padding: 10px;
  border-radius: 8px 0;
  margin-top: 15px;
  box-shadow: 0 0 5px 2px ${colorWhite};

  &::before {
    content: 'ERROR:';
    font-size: 0.7rem;
    font-family: ${fontTitleAlt};
    color: ${colorGrayNormal};
  }

  @media all and (min-width: 768px) {
    width: 60vw;
    padding: 10px 40px;
  }
`;
ErrorDetailsArea.displayName = 'ErrorDetailsArea';

export const ErrorSection = styled.div`
  ${MainBackground}
  height: ${`${window.innerHeight}px`};
  padding: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;

  @media all and (min-width: 768px) {
    ${FullScreen};
    display: block;
  }
`;
ErrorSection.displayName = 'ErrorSection';

export const ErrorTitle = styled.h2`
  margin-left: 0;
  width: auto;
  font-family: ${fontSubtitle};
  font-size: 2rem;
  text-align: left;
`;
ErrorTitle.displayName = 'ErrorTitle';

export const NotificationArea = styled.div`
  position: relative;
  top: 5%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 80%;

  @media all and (min-width: 768px) {
    top: 50%;
    width: 80vw;
  }
`;
NotificationArea.displayName = 'NotificationArea';

interface NotifyButtonProps extends React.ComponentPropsWithoutRef<'a'> {
  visible?: boolean | null;
}

export const NotifyButton = styled(LinkButton)<NotifyButtonProps>`
  margin-top: 40px;
  position: fixed;
  bottom: 5%;
`;
NotifyButton.displayName = 'NotifyButton';

export const NotifyButtonIcon = styled.i`
  display: inline;
  margin-left: 15px;
  vertical-align: middle;
  text-shadow: 0 0 1px ${colorBlack};
  font-size: 2rem;
`;
NotifyButtonIcon.displayName = 'NotifyButtonIcon';

export const NotifyButtonText = styled.p`
  position: relative;
  right: 0;
  margin: 4px 0 0 5px;
  font-family: ${fontTitleAlt};
  font-size: 1.3rem;
  color: ${colorBlack};
`;
NotifyButtonText.displayName = 'NotifyButtonText';

export const SorryText = styled.p`
  margin: 0;
  width: 100%;
  font-style: italic;
  font-family: ${fontSubtitle};
  font-size: 1.1rem;
  line-height: 1.5;
  color: ${colorBlueLight};
  text-align: left;

  @media all and (min-width: 768px) {
    font-size: 1.3rem;
    margin: 0 0 20px;
    width: 80%;
  }
`;
SorryText.displayName = 'SorryText';
