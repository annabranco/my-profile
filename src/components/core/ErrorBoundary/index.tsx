import { Component, ErrorInfo, ReactNode } from 'react';
import { connect } from 'react-redux';
import { node } from 'prop-types';
import { IAppState, IErrorText } from '../../../types/interfaces';
import ErrorComponent from './ErrorComponent';
import { errorTextsPropType } from '../../../types/propTypes';
import defaultTexts from '../../../db/texts.json';

interface ErrorTexts {
  texts: IErrorText | undefined;
}

interface Props {
  children: ReactNode | ReactNode[];
  texts: IErrorText | undefined;
}

interface State {
  error: Error | null;
}

class ErrorBoundary extends Component<Props, State> {
  static propTypes = {
    children: node.isRequired,
    texts: errorTextsPropType
  };

  static defaultProps = {
    texts: defaultTexts.en.error
  };

  state: State = {
    error: null
  };

  static getDerivedStateFromError(error: Error): { error: Error } {
    return { error };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error('Something unexpected had happened', error, errorInfo);
  }

  render() {
    const { error }: State = this.state;
    const { texts }: ErrorTexts = this.props;

    if (error) {
      return <ErrorComponent error={error.message} texts={texts} />;
    }
    return this.props.children;
  }
}

const mapStateToProps = (state: IAppState): ErrorTexts => ({
  texts: state.texts.error
});

export default connect(mapStateToProps)(ErrorBoundary);
