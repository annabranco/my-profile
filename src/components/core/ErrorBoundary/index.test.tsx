import { shallow, ShallowWrapper } from 'enzyme';
import 'jest-styled-components';
import ErrorBoundary from '.';
import { setupMockProvider } from '../../../tests/mocks';

const MockChild = () => <p>Mock Child</p>;

describe('< ErrorBoundary >', () => {
  let MockProvider;
  let wrapper: ShallowWrapper;

  beforeAll(() => {
    MockProvider = setupMockProvider({ language: 'en' }).MockProvider;
    wrapper = shallow(
      <MockProvider>
        <ErrorBoundary>
          <MockChild />
        </ErrorBoundary>
      </MockProvider>
    );
  });

  it('should mount', () => {
    expect(wrapper).toBeDefined();
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  xit('should pass caught error to an Error Component', () => {});
});
