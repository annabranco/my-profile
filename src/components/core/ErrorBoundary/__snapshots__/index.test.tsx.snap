// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< ErrorBoundary > matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ErrorBoundary)>
    <MockChild />
  </Connect(ErrorBoundary)>
</Provider>
`;
